src/components/exercise/index.cjsx

```diff
@@ -1,131 +1,160 @@
 React = require 'react'
 _ = require 'underscore'
 
-camelCase = require 'camelcase'
+ExercisePart = require './part'
+{ExFooter} = require './controls'
+{CardBody} = require '../pinned-header-footer-card/sections'
+ExerciseGroup = require './group'
 
-ExerciseStepCard = require './step-card'
-{propTypes} = require './props'
-{step} = propTypes.ExerciseStepCard
+{ScrollListenerMixin} = require 'react-scroll-components'
+{ScrollTracker, ScrollTrackerParentMixin} = require '../scroll-tracker'
 
-# TODO clean this up.
-REVIEW_CONTROL_PROPS = ['refreshStep', 'recoverFor', 'canTryAnother']
+ExerciseMixin =
+  getLastPartId: ->
+    {parts} = @props
+    _.last(parts).id
 
-NOT_REVIEW_PROPS = ['onNextStep', 'canReview', 'disabled']
-NOT_TEACHER_READ_ONLY_PROPS = _.union(NOT_REVIEW_PROPS, ['onStepCompleted', 'canTryAnother'])
-NOT_MULTIPLE_CHOICE_PROPS = _.union(REVIEW_CONTROL_PROPS, ['disabled'])
-NOT_FREE_RESPONSE_PROPS = _.union(REVIEW_CONTROL_PROPS, ['onStepCompleted', 'onNextStep', 'canReview'])
+  isSinglePart: ->
+    {parts} = @props
+    parts.length is 1
+
+  isLastPart: (id) ->
+    lastPartId = @getLastPartId()
+    lastPartId is id
 
-Exercise = React.createClass
-  displayName: 'Exercise'
-  propTypes:
-    id: React.PropTypes.string.isRequired
-    taskId: React.PropTypes.string.isRequired
-    onStepCompleted: React.PropTypes.func.isRequired
-    onNextStep: React.PropTypes.func.isRequired
-    getCurrentPanel: React.PropTypes.func.isRequired
-    step: step
+  canAllContinue: ->
+    {parts, canOnlyContinue} = @props
+
+    _.reduce parts, (previous, part) ->
+      previous and canOnlyContinue(part.id)
+    , true
+
+  shouldControl: (id) ->
+    {canOnlyContinue} = @props
+
+    not (@isLastPart(id) and canOnlyContinue(id))
+
+  renderPart: (part, partProps) ->
+    props = _.omit(@props, 'part', 'canOnlyContinue', 'footer', 'setScrollState', 'goToStep')
+
+    <ExercisePart
+      {...partProps}
+      {...props}
+      step={part}
+      id={part.id}
+      taskId={part.task_id}/>
```

Comment:
`renderPart` is called from a `map` and is missing a 🔑 property.  I noticed that the demo is complaining: `Warning: Each child in an array or iterator should have a unique "key" prop. Check the render method of ExerciseWithScroll.`  I think this is the cause.